### PR TITLE
Reproduce pymol's surface_ramp_above_mode=1

### LINKED ISSUE #1071: Reproduce pymol's surface_ramp_above_mode=1
I want to reproduce in Mol* the effect of `surface_solvent=0` and `surface_ramp_above_mode=1` [in PyMOL](https://pymolwiki.org/index.php/Surface_ramp_above_mode), which is often used when visualizing the electrostatic potential of a protein.
This means that the surface colour is determined by the function value 1.4A (or one solvent radius) away from the surface, rather than the function value at the surface.
Currently there is no Mol* way of doing this.

### PR DESCRIPTION
**Pull request by hui2000ji**
# Description

This is an initial attempt to fix #1071. It is **NOT READY FOR MERGING** yet.

- With this fix enabled, all surface will be coloured by function values 1.4A away from the point on surface ("away" means the direction of the normal vector at this point). This value is now hard-coded. Ideally, the user should be able to enable/disable this behaviour and configure the value 1.4A.
- I only tested on "Molecular Surface Mesh" and did not check the behaviour on the other two types of molecular surface visualization.

image

# How I come up with this fix

Below is a brief description on how creating a Molecular Surface with a UnitsMeshVisual works.

- `mol-repr/structure/visual/molecular-surface-mesh.ts`: This is the entry point of creating a molecular surface. It invokes `UnitsMeshVisual`.
- `mol-repr/structure/units-visual.ts`: Where `UnitsMeshVisual` is defined. It invokes:
    - `positionIt = createPositionIterator(newGeometry, renderObject.values)`
        - Type definitions: Mesh, TransformData -> LocationIterator
        - Implemented in `mol-geo/geometry/mesh/mesh.ts`
    - `createColors(locationIt, positionIt, newTheme.color, renderObject.values)`
        - Type definitions: LocationIterator, LocationIterator, LocationColor, ColorData -> ColorData
        - Implemented in `mol-geo/geometry/color-data.ts`
    - The `color` function, of type LocationColor, is implemented in `mol-theme/color/external-volume.ts`
        - Type definitions: Location -> Color. Note that granularity is "vertex".

Note that `createColors` accept a location iterator and a position iterator. When creating colors when `colorTheme.granularity == 'vertex'` (which is the case for the External Volume theme), only the position iterator is used.
So the fix is done by changing the logic of `createPositionIterator`. Instead of returning the position for each point on the surface, it returns the position of a point that is 1.4A away from (moving by the direction of the normal vector at this point) the point on the surface.

# How to test the effect

For PDB 1n8y, I have calculated an electrostatic potential using APBS, with results stored at [on Google Drive](https://drive.google.com/file/d/19GgEFSW0LvTbfubE2KKODAA1iWsP3T4i/view?usp=sharing).
To test the effect,

- download PDB 1n8y
- open the `map.dx` file, and hide it
image
- create a molecular surface on the 1n8y "Polymer" object with the following configurations.
image

Before the fix, the result would look like this:
image

After the fix, the result would look like this, which is identical to the PyMOL visualization with surface_ramp_above_mode=1.
image

# Actions

- [x] Added description of changes to the `[Unreleased]` section of `CHANGELOG.md`
- [x] Updated headers of modified files
- [x] Added my name to `package.json`'s `contributors`

### CODE REVIEW

**Comment by arose** [comment]
I like this feature! I suggest to add the normal as an optional property to `PositionLocation` and do the normal + position in the color theme. That way you can make the `offset` value part of the color theme params.

**Comment by hui2000ji** [comment]
Hi Alex! Thanks for your enlightening suggestions. I have made the corresponding changes. Please take a look.

**Review by arose — CHANGES_REQUESTED** [review]
Thanks @hui2000ji. I tested this with results from https://server.poissonboltzmann.org/ and it works very ncely. I made a few inline comments, but once those are adressed we can merge it.

File: src/mol-geo/geometry/mesh/mesh.ts [hunk 1]
```diff
@@ -651,13 +651,17 @@ export namespace Mesh {
...
         const m = transform.aTransform.ref.value;
         const getLocation = (groupIndex: number, instanceIndex: number) => {
             if (instanceIndex < 0) {
-                Vec3.fromArray(p, v, groupIndex * 3);
+                Vec3.fromArray(p, vs, groupIndex * 3);
+                Vec3.fromArray(n, ns, groupIndex * 3);
             } else {
-                Vec3.transformMat4Offset(p, v, m, 0, groupIndex * 3, instanceIndex * 16);
+                Vec3.transformMat4Offset(p, vs, m, 0, groupIndex * 3, instanceIndex * 16);
+                Vec3.transformMat4Offset(n, ns, m, 0, groupIndex * 3, instanceIndex * 16);
```
[the remark this answers is not in the document]
Comment (hui2000ji) [reply src/mol-geo/geometry/mesh/mesh.ts:664]: Should I calculate the transpose of the inverse within this new method?

```typescript
/**
 * Transforms the direction vector with a Mat4. 4th vector component is implicitly '0'
 * This means the translation components of the matrix are ignored.
 * This function is useful for transforming normal vectors with non-uniform scaling.
 */
export function transformDirectionOffset(out: NumberArray, a: NumberArray, m: NumberArray, outO: number, aO: number, oM: number) {
    // Assumes that m is already the transpose of the inverse matrix suitable for normal transformation.
    const x = a[0 + aO], y = a[1 + aO], z = a[2 + aO];
    out[0 + outO] = m[0 + oM] * x + m[4 + oM] * y + m[8 + oM] * z;
    out[1 + outO] = m[1 + oM] * x + m[5 + oM] * y + m[9 + oM] * z;
    out[2 + outO] = m[2 + oM] * x + m[6 + oM] * y + m[10 + oM] * z;
    // Normalize the output vector to handle non-uniform scaling
    const len = Math.hypot(out[0 + outO], out[1 + outO], out[2 + outO]);
    if (len > 0) {
        out[0 + outO] /= len;
        out[1 + outO] /= len;
        out[2 + outO] /= len;
    }
    return out;
}
```

[the remark this answers is not in the document]
Comment (arose) [reply src/mol-geo/geometry/mesh/mesh.ts:664]: > Should I calculate the transpose of the inverse within this new method?

No, let's assume the transformation matrix is "well behaved". I will make an issue to pre-calculate the transpose of the inverse so it is available in the `TransformData` and we can use it here and also in the `mesh.vert.glsl` shader.

**Comment by hui2000ji** [comment]
Hi @arose, I tried to address your comments. Please check if everything's good now.

**Review by arose — APPROVED** [review]

**Comment by arose** [comment]
Thank you @hui2000ji, very nicely done!

**Comment by dsehnal** [comment]
Good job @hui2000ji !

**Pull request merged**